### PR TITLE
[cluster_test] Add addl params to HealthCheckRunner::run()

### LINKED ISSUE #1244: [cluster_test] Do not print health checks when everything is ok
Currently when experiment runs cluster-test output is very verbose, mostly because `HealthCheckRunner::run` prints a lot of stuff on every tick.
Instead of printing it every time, `run` should probably take boolean arg like `print_on_failure` and only print output when this healthcheck failed.

One caveat is that health checker is currently not aware of what validators are affected by experiment. It instead returns list of failed validators and calling code can bi-sect it with list of validators under experiment, in order to make decision on whether test failed or not.

https://github.com/libra/libra/blob/e98af2dae812b7da1024d215321ab7d05aa5522c/testsuite/cluster-test/src/main.rs#L446

One solution is instead of returning list of failed validators, just pass list of validators under experiment into health check, and make HealthCheck::run returning `failure::Result<()>`

### PR DESCRIPTION
**Pull request by ankushagarwal**
# Summary
HealthCheckRunner::run() now takes an additional parameter
affected_validators_set. If there are validators which failed which were not
part of the experiment, then HealthCheckRunner::run()
returns an Err with a string of all the unexpected
failures. Otherwise, it returns a list of ALL the failed validators.

HealthCheckRunner::run() now also takes a bool parameter: only_print_on_failure.
If this is set to true, messages are printed only when there are
failures. If this is set to false, messages are always printed

Closes #1244

## Test Plan

Untested. Unable to run cluster-test on AWS.

### CODE REVIEW

File: testsuite/cluster-test/src/health/mod.rs [hunk 1]
```diff
@@ -93,12 +96,24 @@ impl HealthCheckRunner {
...
+    /// of all the unexpected failures.
+    /// Otherwise, it returns a list of ALL the failed validators
+    /// It also takes a bool parameter: only_print_on_failure. If this is set
+    /// to true, messages are printed only when there are failures.
+    /// If this is set to false, messages are always printed
+    pub fn run(
+        &mut self,
+        events: &[ValidatorEvent],
+        affected_validators_set: &HashSet<String>,
+        only_print_on_failure: bool,
```
Comment (ankushagarwal) [inline testsuite/cluster-test/src/health/mod.rs]: @andll : I have set this to true at all call sites. I am not sure at which call sites this should be set to true and where it should be set to false.

File: testsuite/cluster-test/src/health/mod.rs [hunk 2]
```diff
@@ -93,12 +96,24 @@ impl HealthCheckRunner {
...
+    ) -> failure::Result<Vec<String>> {
```
Comment (ankushagarwal) [inline testsuite/cluster-test/src/health/mod.rs]: @andll In the issue, you suggested a return type of `failure::Result<()>`, but there are clients which use the list of failed nodes. Example: `cluster-test/src/main.rs:497`. So I used return type of `failure::Result<Vec<String>>`

File: testsuite/cluster-test/src/main.rs [hunk 3]
```diff
@@ -512,7 +510,7 @@ impl ClusterTestRunner {
             // This assumes so far that event propagation time is << 1s, this need to be refined
             // in future to account for actual event propagation delay
             let events = self.logs.recv_all_until_deadline(deadline);
-            self.health_check_runner.run(&events);
+            self.health_check_runner.run(&events, &HashSet::new(), true);
```
Comment (ankushagarwal) [inline testsuite/cluster-test/src/main.rs]: Here compiler gives a warning that we are not using `std::result::Result`. 

I guess we don't care about the returned Result here, we only want to use this to print healthcheck messages.

File: testsuite/cluster-test/src/health/mod.rs (hunk 1 above)
Comment (andll) [reply testsuite/cluster-test/src/health/mod.rs]: Basically when you run cluster test with `--run-once` and if test passes, we don't want to see those healthchecks

File: testsuite/cluster-test/src/health/mod.rs (hunk 2 above)
Comment (andll) [reply testsuite/cluster-test/src/health/mod.rs]: `cluster-test/src/main.rs:497` seem to be something else now, can you give code snippet?

File: testsuite/cluster-test/src/main.rs (hunk 3 above)
Comment (andll) [reply testsuite/cluster-test/src/main.rs]: Yeah, you can do something like `let _ignore = self. health_check_runner.run...`

File: testsuite/cluster-test/src/health/mod.rs (hunk 2 above)
Comment (ankushagarwal) [reply testsuite/cluster-test/src/health/mod.rs]: Here the return value from HealthCheckRunner::run() is being use to break from the loop. This is in `fn run_single_experiment` in `cluster-test/src/main.rs`

```
            let unhealthy_validators;
            match self
                .health_check_runner
                .run(&events, &affected_validators, true)
            {
                Err(s) => bail!("Validators which were not under experiment failed : {}", s),
                Ok(r) => unhealthy_validators = r,
            }
            if unhealthy_validators.is_empty() {
                break;
            }
```

File: testsuite/cluster-test/src/main.rs (hunk 3 above)
Comment (ankushagarwal) [reply testsuite/cluster-test/src/main.rs]: Updated

File: testsuite/cluster-test/src/health/mod.rs (hunk 1 above)
Comment (ankushagarwal) [reply testsuite/cluster-test/src/health/mod.rs]: Okay. Is there any scenario where we want to print health-checks even when the health-checks are passing?

If there isn't such a scenario, we can eliminate this parameter and only print health-checks when there are failures.

**Comment by ankushagarwal** [comment]
@andll : I am not sure how to test this right now because I have issues with running cluster-test in my workspace..

File: testsuite/cluster-test/src/health/mod.rs (hunk 2 above)
Comment (andll) [reply testsuite/cluster-test/src/health/mod.rs]: Looks like it uses `unhealthy_validators` to generate error message, so we should be able to just reuse error message right?

**Comment by andll** [comment]
@ankushagarwal what tag did you use to deploy your workplace? You should be able to use `nightly_tested` image to run cluster

File: testsuite/cluster-test/src/health/mod.rs (hunk 1 above)
Comment (andll) [reply testsuite/cluster-test/src/health/mod.rs]: We want health check to still print on `--healthcheck`...

Comment (andll) [reply testsuite/cluster-test/src/health/mod.rs]: Thats actually probably only place where we want them all the time

**Comment by ankushagarwal** [comment]
I used the default values from `facebook/dev.tfvars` which had the nightly tag.

I just updated to use `nightly_tested` using `terraform apply --var-file=facebook/dev.tfvars -var "image_tag=nightly_tested"`.

Let me check if that works.

**Comment by ankushagarwal** [comment]
Updated so that:

1. healthcheck is always printed when action is  `--health-check `
2. healthcheck is only printed on failures for every other scenario.

Was finally able to get my test cluster up and running and tested this by running the `--health-check ` action and the `--run-once` action.

File: testsuite/cluster-test/src/health/mod.rs (hunk 2 above)
Comment (ankushagarwal) [reply testsuite/cluster-test/src/health/mod.rs]: It is using the result (`unhealthy_validators`) to determine whether to break from the loop, not just for the error message.

File: testsuite/cluster-test/src/health/mod.rs (hunk 1 above)
Comment (ankushagarwal) [reply testsuite/cluster-test/src/health/mod.rs]: Updated.

**Comment by andll** [comment]
I think there are some build failures still - could you fix them?

File: testsuite/cluster-test/src/health/mod.rs (hunk 2 above)
Comment (andll) [reply testsuite/cluster-test/src/health/mod.rs]: But criteria on whether to break the loop is same as criteria inside health check on whether to print error?
So we should not duplicate it...

**Comment by ankushagarwal** [comment]
@andll Fixed CI failures. Please take a look.

**Pull request closed without merging**